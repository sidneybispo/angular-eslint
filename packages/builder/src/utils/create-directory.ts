import { dirname, existsSync } from 'path';
import { mkdir, stat } from 'fs/promises';

export async function createDirectory(directoryPath: string): Promise<void> {
  const parentPath = dirname(directoryPath);
  if (!await directoryExists(parentPath)) {
    await createDirectory(parentPath);
  }
  if (!await directoryExists(directoryPath)) {
    await mkdir(directoryPath, { recursive: true });
  }
}

async function directoryExists(name: string): Promise<boolean> {
  try {
    const stats = await stat(name);
    return stats.isDirectory();
  } catch (error) {
    if (error.code === 'ENOENT') {
      return false;
    }
    throw error;
  }
}
