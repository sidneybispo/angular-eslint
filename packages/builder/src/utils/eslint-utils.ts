import * as ESLintLibrary from 'eslint';
import { join } from 'path';
import { readFileSync } from 'fs';

export async function loadESLint(): Promise<typeof ESLintLibrary> {
  let eslint;
  try {
    // Use dynamic import here to avoid importing ESLint unnecessarily
    eslint = await import('eslint');
    return eslint;
  } catch (error) {
    throw new Error('Unable to find ESLint. Ensure ESLint is installed: ' + error.message);
  }
}

export async function lint(
  workspaceRoot: string,
  eslintConfigPath: string | undefined,
  options: ESLintLibrary.ESLint.Options & { lintFilePatterns: string[] },
): Promise<ESLintLibrary.ESLint.LintResult[]> {
  const projectESLint = await loadESLint();

  const eslintConfig = eslintConfigPath
    ? JSON.parse(readFileSync(eslintConfigPath, 'utf-8'))
    : undefined;

  const eslint = new projectESLint.ESLint({
    ...(eslintConfig && { overrideConfigFile: eslintConfig }),
    ignorePath: options.ignorePath,
    fix: options.fix,
    cache: options.cache,
    cacheLocation: options.cacheLocation,
    cacheStrategy: options.cacheStrategy,
    resolvePluginsRelativeTo: options.resolvePluginsRelativeTo,
    rulePaths: options.rulePaths,
    errorOnUnmatchedPattern: false,
  });

  return await eslint.lintFiles(
    options.lintFilePatterns.map((p) => join(workspaceRoot, p)),
  );
}
