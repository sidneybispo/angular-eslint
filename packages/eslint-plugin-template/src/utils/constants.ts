/**
 * A symbol representing a property read.
 * @example
 * ```html
 * <input [disabled]="disabled">
 * ```
 */
export const propertyRead = Symbol('propertyRead');
