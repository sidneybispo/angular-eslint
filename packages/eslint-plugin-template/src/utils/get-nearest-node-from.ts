import type { AST, R3_Node as Node } from '@angular-eslint/bundled-angular-compiler';
import type { TSESTree } from '@typescript-eslint/experimental-utils';
import { AST_NODE_TYPES } from '@typescript-eslint/types';

type ASTOrNodeWithParent = (AST | Node) & { parent?: ASTOrNodeWithParent };

function isProgram(node: unknown): node is TSESTree.Program {
  return (node as { type?: string }).type === AST_NODE_TYPES.Program;
}

function getNearestParent<T extends ASTOrNodeWithParent>(
  node: ASTOrNodeWithParent,
  predicate: (parent: ASTOrNodeWithParent) => parent is T,
): T | null {
  let currentNode: ASTOrNodeWithParent | null = node.parent;

  while (currentNode && !isProgram(currentNode)) {
    if (predicate(currentNode)) {
      return currentNode;
    }

    currentNode = currentNode.parent;
  }

  return null;
}

// Usage example:
const nearestTemplateNode = getNearestParent(templateNode, (node): node is Node.Template => node.kind === 'template');
