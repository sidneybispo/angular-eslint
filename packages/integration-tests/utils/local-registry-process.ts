import type { ChildProcess } from 'child_process';
import { spawn } from 'child_process';
import execa from 'execa';
import { join } from 'path';
import kill from 'tree-kill';
import type { ExecaChildProcess } from 'execa';

export const FIXTURES_DIR = join(__dirname, '../fixtures/');
export const LONG_TIMEOUT_MS = 600000; // 10 mins

let localRegistryProcess: ChildProcess | null = null;

const VERSION = `9999.0.1-local-integration-tests`;
const cwd = process.cwd();

const setupLocalRegistry = async (): Promise<void> => {
  try {
    process.env.npm_config_registry = `http://localhost:4872/`;
    process.env.YARN_REGISTRY = process.env.npm_config_registry;

    localRegistryProcess = spawn('npx', [
      'verdaccio',
      '--config',
      './local-registry/config.yml',
      '--listen',
      '4872',
    ]);

    return new Promise((resolve, reject) => {
      const collectedOutput: string[] = [];

      localRegistryProcess.stdout?.on('data', (data: Buffer) => {
        collectedOutput.push(data.toString());
        if (data.toString().includes('http address')) {
          resolve(undefined);
        }
      });

      localRegistryProcess.stderr?.on('data', (data: Buffer) => {
        collectedOutput.push(data.toString());
      });

      localRegistryProcess.on('error', (err: Error) => {
        console.error(collectedOutput.join(''));
        reject(err);
      });

      localRegistryProcess.on('exit', (code) => {
        if (code !== 0) {
          console.error(`Failed to start the npm registry: exit code ${code}`);
          console.error(collectedOutput.join(''));
          reject(new Error(`Failed to start the npm registry`));
        }
      });
    });
  } catch (err) {
    console.error(`Failed to start the npm registry`, err);
    throw err;
  }
};

const cleanup = (code: number): void => {
  if (localRegistryProcess) localRegistryProcess.kill();

  try {
    kill(0);
    kill(0, 'SIGKILL');
    // eslint-disable-next-line no-empty
  } catch (err) {}

  process.chdir(cwd);
  process.exit(code);
};

const checkLocalRegistry = (): void => {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }
};

export { setupLocalRegistry, cleanup, checkLocalRegistry };

export const publishPackagesToVerdaccio = async (): Promise<ExecaChildProcess<string>> => {
  checkLocalRegistry();

  const subprocess = execa('./publish-to-verdaccio.sh', [VERSION]);
  subprocess.stdout?.pipe(process.stdout);
  subprocess.stderr?.pipe(process.stderr);

  return await subprocess;
};

export const runNpmInstall = async (): Promise<ExecaChildProcess<string>> => {
  checkLocalRegistry();

  const subprocess = execa('npm', ['install']);
  subprocess.stdout?.pipe(process.stdout);
  subprocess.stderr?.pipe(process.stderr);

  return await subprocess;
};

// ... Add other functions here, following the same pattern
