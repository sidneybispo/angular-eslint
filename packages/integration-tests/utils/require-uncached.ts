/**
 * It's important in our snapshot assertions which target modules on disk
 * that we are always grabbing a fresh copy of the file. Node's built-in
 * caching mechanism can get in the way of that so this utility helps
 * us work around the potential issue.
 *
 * (Having the return type as any makes it much easier to work with).
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function requireUncached(modulePath: string): any {
  const moduleKey = require.resolve(modulePath);
  if (moduleKey in require.cache) {
    delete require.cache[moduleKey];
  }
  return require(modulePath);
}
