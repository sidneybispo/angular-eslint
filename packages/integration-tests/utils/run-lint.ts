import execa from 'execa';
import path from 'path';
import stripAnsi from 'strip-ansi';

const FIXTURES_DIR = path.join(__dirname, '../fixtures/');

type NormalizeOutputFn = (value: string) => string;

const normalizeOutput: NormalizeOutputFn = (value) => {
  const ansiRemoved = stripAnsi(value);
  return ansiRemoved.replace(
    new RegExp(`^${FIXTURES_DIR.replace(/\\/g, '\\\\')}(.*?)$`, 'gm'),
    (_, c1) => `__ROOT__/${c1.replace(/\\/g, '/')}`,
  );
};

type RunLintFn = (directory: string) => Promise<string | undefined>;

const runLint: RunLintFn = async (directory) => {
  try {
    const command = 'npx';
    const args = ['ng', 'lint'];
    const cwd = path.join(FIXTURES_DIR, directory);

    const subprocess = execa(command, args, { cwd });

    subprocess.stdout?.pipe(process.stdout);
    subprocess.stderr?.pipe(process.stderr);

    const { stdout } = await subprocess;

    return normalizeOutput(stdout);
  } catch (error) {
    if (error.code === 'ENOENT') {
      console.error(`Command '${error.command}' not found.`);
    } else {
      console.error(error);
    }
  }
};

export default runLint;
