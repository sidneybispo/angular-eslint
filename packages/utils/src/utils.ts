/**
 * General-purpose utilities which are not specific to one of the plugins.
 */

/**
 * Returns the last item of the given array.
 */
export function getLast<T extends readonly unknown[]>(items: T): T[number] {
  return items[items.length - 1];
}

export const objectKeys = Object.keys as <T>(o: T) => readonly Extract<keyof T, string>[];

/**
 * Enforces the invariant that the input is an array.
 */
export function arrayify<T>(value: T | readonly T[]): readonly T[] {
  return Array.isArray(value) ? value : (value ? [value] : []);
}

/**
 * Checks if the input is not null or undefined.
 */
export const isNotNullOrUndefined = <T>(input: null | undefined | T): input is T => input != null;

export const kebabToCamelCase = (value: string): string =>
  value.replace(/-[a-zA-Z]/g, (match) => match[1].toUpperCase());

/**
 * Converts an array to human-readable text.
 */
export const toHumanReadableText = (items: readonly string[]): string => {
  const itemsLength = items.length;

  if (itemsLength === 1) {
    return `"${items[0]}"`;
  }

  return `${items.slice(0, -1).join(', ')} or "${items[itemsLength - 1]}"`;
};

export const toPattern = (value: readonly unknown[]): RegExp =>
  new RegExp(`^(${value.join('|')})$`);

/**
 * Capitalizes the first letter of the given text.
 */
export function capitalize<T extends string>(text: T): Capitalize<T> {
  return text.charAt(0).toUpperCase() + text.slice(1) as Capitalize<T>;
}

/**
 * Removes brackets and whitespaces from the given text.
 */
export function withoutBracketsAndWhitespaces(text: string): string {
  return text.replace(/[\[\]\s]/g, '');
}
