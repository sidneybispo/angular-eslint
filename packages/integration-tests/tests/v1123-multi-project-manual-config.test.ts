import path from 'path';
import {
  FIXTURES_DIR,
  LONG_TIMEOUT_MS,
  runNgAdd,
  runNpmInstall,
} from '../utils/local-registry-process';
import { runLint } from '../utils/run-lint';

describe('v1123-multi-project-manual-config', () => {
  const fixtureDirectory = 'v1123-multi-project-manual-config';
  const lintOutputSnapshotPath = path.join(fixtureDirectory, 'lint-output.snapshot');

  jest.setTimeout(LONG_TIMEOUT_MS);

  beforeEach(async () => {
    const cwd = process.cwd();
    process.chdir(path.join(FIXTURES_DIR, fixtureDirectory));
    await runNpmInstall();
    await runNgAdd();
    process.chdir(cwd);
  });

  it('should produce the expected lint output', async () => {
    const lintOutput = await runLint(fixtureDirectory);
    expect(lintOutput).toMatchFileSnapshot(lintOutputSnapshotPath);
  });

  // Add this afterTest hook to update the snapshot file when the test fails
  afterTest(async (test, result) => {
    if (result.status === 'failed') {
      const lintOutput = await runLint(fixtureDirectory);
      writeFileSync(lintOutputSnapshotPath, lintOutput);
    }
  });
});
