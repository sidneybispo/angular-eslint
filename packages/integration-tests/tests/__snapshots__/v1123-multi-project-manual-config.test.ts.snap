// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lint test for v1123-multi-project-manual-config`] = `
Linting "v1123-multi-project-manual-config"...

${Array.from({ length: 20 }, (_, i) => {
  return `
  __ROOT__/v1123-multi-project-manual-config/src/app/app.component.html
   1:31  error  Invalid binding syntax. Use [(expr)] instead  @angular-eslint/template/banana-in-box
`;
}).join('')}

${Array.from({ length: 1 }, (_, i) => {
  return `
  __ROOT__/v1123-multi-project-manual-config/src/app/app.component.spec.ts
   19:6  error  Strings must use singlequote  @typescript-eslint/quotes
`;
}).join('')}

${Array.from({ length: 1 }, (_, i) => {
  return `
  __ROOT__/v1123-multi-project-manual-config/src/app/app.component.ts
   2:1  error  'rxjs/Rx' import is restricted from being used. Please import directly from 'rxjs' instead                             no-restricted-imports
   5:1  error  Unexpected console statement                                                                                           no-console
  15:3  error  Lifecycle interface 'OnInit' should be implemented for method 'ngOnInit'. (https://angular.io/styleguide#style-09-01)  @angular-eslint/use-lifecycle-interface
`;
}).join('')}

...

${Array.from({ length: 1 }, (_, i) => {
  return `
  __ROOT__/v1123-multi-project-manual-config/projects/websites-system/src/test.ts
   13:5  error  Member call should be declared before all method definitions  @typescript-eslint/member-ordering
`;
}).join('')}

${Array.from({ length: 1 }, (_, i) => {
  return `
  __ROOT__/v1123-multi-project-manual-config/projects/robot-websites/src/test.ts
   13:5  error  Member call should be declared before all method definitions  @typescript-eslint/member-ordering
`;
}).join('')}


\`\`\`
âœ– ${20 + 43 + 1 + 1 + 1 + 1 + 1 + 1} problems (... errors, 0 warnings)
  ${12} errors and 0 warnings potentially fixable with the \`--fix\` option.
\`\`\`

`;

